lighthouse-core/audits/byte-efficiency/render-blocking-resources.js

```diff
@@ -0,0 +1,188 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+/**
+ * @fileoverview Audit a page to see if it does have resources that are blocking first paint
+ */
+
+'use strict';
+
+const Audit = require('../audit');
+const Node = require('../../lib/dependency-graph/node');
+const ByteEfficiencyAudit = require('./byte-efficiency-audit');
+const UnusedCSS = require('./unused-css-rules');
+const WebInspector = require('../../lib/web-inspector');
+
+// Because of the way we detect blocking stylesheets, asynchronously loaded
+// CSS with link[rel=preload] and an onload handler (see https://github.com/filamentgroup/loadCSS)
+// can be falsely flagged as blocking. Therefore, ignore stylesheets that loaded fast enough
+// to possibly be non-blocking (and they have minimal impact anyway).
+const MINIMUM_WASTED_MS = 50;
+
+const keyByUrl = arr => arr.reduce((map, node) => {
+  map[node.record && node.record.url] = node;
+  return map;
+}, {});
+
+class RenderBlockingResources extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      name: 'render-blocking-resources',
+      description: 'Eliminate render-blocking resources',
+      informative: true,
+      scoreDisplayMode: Audit.SCORING_MODES.NUMERIC,
+      helpText: 'Resources are blocking the first paint of your page. Consider ' +
+          'delivering critical JS/CSS inline and deferring all non-critical ' +
+          'JS/styles. [Learn more](https://developers.google.com/web/tools/lighthouse/audits/blocking-resources).',
+      requiredArtifacts: ['CSSUsage', 'URL', 'TagsBlockingFirstPaint', 'traces'],
+    };
+  }
+
+  /**
+   * @param {Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   */
+  static async computeResults(artifacts, context) {
+    const trace = artifacts.traces[Audit.DEFAULT_PASS];
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const simulatorData = {devtoolsLog, settings: context.settings};
+    const traceOfTab = await artifacts.requestTraceOfTab(trace);
+    const simulator = await artifacts.requestLoadSimulator(simulatorData);
+    const wastedBytesMap = await RenderBlockingResources.computeWastedCSSBytes(artifacts, context);
+
+    const metricSettings = {throttlingMethod: 'simulate'};
+    const metricComputationData = {trace, devtoolsLog, simulator, settings: metricSettings};
+    const fcpSimulation = await artifacts.requestFirstContentfulPaint(metricComputationData);
+    const fcpTsInMs = traceOfTab.timestamps.firstContentfulPaint / 1000;
+
+    const nodeTimingMap = fcpSimulation.optimisticEstimate.nodeTiming;
+    const nodesByUrl = keyByUrl(Array.from(nodeTimingMap.keys()));
```

Comment:
can we instead just have a helper function of like `getNodeTimingForUrl(nodeTimingMap, url){ ..`

Reply:
and the helper function loops through the keys to find the matching record?